openraft/src/engine.rs

```diff
@@ -0,0 +1,259 @@
+use std::collections::BTreeSet;
+use std::ops::Range;
+use std::sync::Arc;
+
+use maplit::btreeset;
+
+use crate::entry::RaftEntry;
+use crate::error::InitializeError;
+use crate::error::NotAllowed;
+use crate::storage::InitialState;
+use crate::EffectiveMembership;
+use crate::LogId;
+use crate::LogIdOptionExt;
+use crate::Membership;
+use crate::MetricsChangeFlags;
+use crate::NodeId;
+use crate::Vote;
+
+/// Raft protocol algorithm.
+///
+/// It implement the complete raft algorithm except does not actually update any states.
+/// But instead, it output commands to let a `RaftRuntime` implementation execute them to actually update the states
+/// such as append-log or save-vote by execute .
+///
+/// This structure only contains necessary information to run raft algorithm,
+/// but none of the application specific data.
+/// TODO: make the fields private
+#[derive(Debug, Clone, Default)]
+pub(crate) struct Engine<NID: NodeId> {
+    /// TODO:
+    #[allow(dead_code)]
+    pub(crate) id: NID,
+
+    /// Cache of a cluster of only this node.
+    ///
+    /// It is used to check an early commit if there is only one node in a cluster.
+    pub(crate) single_node_cluster: BTreeSet<NID>,
```

Comment:
why this field named with `single`, what does `single` mean?

Reply:
It's a cache of a cluster config of only one node.
And it is used only when checking if replication is required:
A single node cluster does not need to replicate log entries. It is able to commit at once a log entry is appended to local store.